File: lib/lotd_web/channels/public_channel.ex
defmodule LotdWeb.PublicChannel do
  use LotdWeb, :channel

  alias Phoenix.View
  alias Lotd.Museum

  def join("public", _params, socket) do

    displays = Museum.list_displays()
    items = Museum.list_items()
    locations = Museum.list_locations()
    quests = Museum.list_quests()
    mods = Museum.list_mods()

    {:ok, %{
      admin: admin?(socket),
      moderator: moderator?(socket),
      user: authenticated?(socket) && socket.assigns.user.id,
      displays: View.render_many(displays, DataView, "display.json" ),
      items: View.render_many(items, DataView, "item.json" ),
      locations: View.render_many(locations, DataView, "location.json" ),
      quests: View.render_many(quests, DataView, "quest.json" ),
      mods: View.render_many(mods, DataView, "mod.json" )
    }, socket}
  end
end

File: assets/js/api/public_channel.js
import $ from 'jquery'
import { socket, join_moderator_channel, join_user_channel } from '.'
import { Data, Table, Flash } from '../utils'

const configure_public_channel = () => {
  let channel = socket.channel("public")

  // listen for events
  channel.on('add-item', ({ item }) => {
    window.item_table.row.add(item).draw().node()
    calculate_item_counts()
  })

  channel.on('delete-item', ({ id }) => {
    window.item_table.row(`#${id}`).remove().draw()
    calculate_item_counts()
  })

  channel.join()
    .receive("ok", params => {

      // user has joined already, do nothing...
      if (window.user != undefined) return

      const { displays, items, locations, mods, quests, user, moderator, admin } = params

      window.user = user
      window.moderator = moderator
      window.admin = admin

      Table.location(locations)
      Table.quest(quests)
      Table.display(displays)
      Table.mod(mods)
      Table.item(items)

      if (user) join_user_channel(user)
      else Data.calculate_item_counts()

      if (moderator) join_moderator_channel()

      $('#loader-wrapper').addClass('d-none')
    })
    .receive("error", ({reason}) => Flash.error(reason))
}

export default configure_public_channel
